guardian/prosemirror-elements
# Update inner selection when outer selection updates selection within a text field

## Background
In prosemirror-elements, elements are a collection of fields representing user-editable data. Most logic managing these fields is in the relevant `FieldView`. Text fields are all based on `ProseMirrorFieldView`, via implementations that extend it, e.g. `TextFieldView`. These contain instances of ProseMirror ('inner editors'), and the 'outer editor' (the editor that contains the element) delegates control of certain actions to the 'inner editor' managed by the FieldView.

Currently, two kinds of changes from the outer editor can cause changes in the field view:
  1. Changes to the nodes that the inner editor manages:
  2. Changes to `Decorations` that affect the range covered by the field view.

However, if the selection is updated by the outer editor, the selection within a relevant field is not updated.

This hasn't been obvious in the past, because usually the selection is updated by the field's 'inner' editor. With notes being supported in nested element fields, when a user hits 'F10' in Composer with text selected inside the nested element, the selection is updated by the outer editor (to which the menu belongs), and should be moved to the end of the note as users expect. Because the selection is not passed to the inner editor, the note content remains selected, so when the user types, they overwrite the content of the note.


## What does this change?

This PR updates field views when the selection changes, and the text fields (extensions of `ProsemirrorFieldView`) update their inner editor selection to match the incoming selection. This means that the usual F10 behaviour for notes also works inside our test fields.

![Kapture 2024-05-14 at 15 49 59](https://github.com/guardian/prosemirror-elements/assets/34686302/2876d6b3-1f84-4aae-af24-8cb31058ef6c)

This happens as follows:
1. The outer selection changes
2. Update` in `plugin.ts`, a function at the boundary of the plugin, receives state changes from the outer editor, which included the updated `Selection`.
3.  We pass this into the `FieldViews` via some intermediate functions.
4. The only `FieldView` that needs to handle the selection change is the abstract `ProseMirrorFieldView`, which all text `FieldViews` extend.
5. Every text `FieldView` checks if the selection covers its own range. If it does, it updates its internal selection.

## How to test
1. Use prosemirror-elements locally with flexible-content using `yalc`.
6. Create a list format document, like Key Takeaways
7. Type some text in the nested element field, highlight some of it, and hit F10. The cursor should move to the end of the note.

**n.b. - due to an unrelated bug with menu clicks causing the editor to lose focus, this only works with the F10 shortcut, not the notes menu button**

I have updated existing tests to match the new update behaviour and added a test to reflect the selection behaviour itself, making sure that the inner selection is properly updated when there is a relevant incoming selection.
